presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -168,6 +168,39 @@ public static Slice stPoint(@SqlType(DOUBLE) double x, @SqlType(DOUBLE) double y
         return serialize(geometry);
     }
 
+    @SqlNullable
+    @Description("Returns a multi-point geometry formed from input points")
+    @ScalarFunction("ST_MultiPoint")
+    @SqlType(GEOMETRY_TYPE_NAME)
+    public static Slice stMultiPoint(@SqlType("array(" + GEOMETRY_TYPE_NAME + ")") Block input)
+    {
+        MultiPoint multipoint = new MultiPoint();
+        for (int i = 0; i < input.getPositionCount(); i++) {
+            // Null point raises exception
+            if (input.isNull(i)) {
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "ST_MultiPoint takes only an array of valid points, NULL point was passed");
+            }
+
+            Slice slice = GEOMETRY.getSlice(input, i);
+            OGCGeometry geometry = deserialize(slice);
+            if (!(geometry instanceof OGCPoint)) {
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("ST_MultiPoint takes only an array of valid points, %s was passed", geometry.geometryType()));
+            }
+            OGCPoint point = (OGCPoint) geometry;
+
+            // Empty points are ignored
+            if (point.isEmpty()) {
```

Comment:
Why are empty points ignored? This doesn't seem to be the behavior described by the spec.

Reply:
@martint It makes no sense to include empty points. Since the spec is not explicitly saying anything about empty points, we are doing what makes most sense.

Reply:
The code for ST_ToPointAry doesn't filter empty points:

```
CREATE FUNCTION ST_ToPointAry (ageometryarray ST_Geometry ARRAY[ST_MaxGeometryArrayElements])
    RETURNS ST_Point ARRAY[ST_MaxGeometryArrayElements]
    LANGUAGE SQL
    DETERMINISTIC
    CONTAINS SQL
    RETURNS NULL ON NULL INPUT
    STATIC DISPATCH
    BEGIN
        DECLARE counter INTEGER;
        DECLARE apointarray ST_Point ARRAY[ST_MaxGeometryArrayElements];
        -- If ageometryarray is the null value or contains null elements,
        -- then raise an exception.
        CALL ST_CheckNulls(ageometryarray);
        -- Set apointarray to an empty array.
        SET apointarray = CAST(ARRAY[] AS ST_Point ARRAY[ST_MaxGeometryArrayElements]);
        SET counter = 1;
        -- For each element in ageometryarray
        WHILE counter <= CARDINALITY(ageometryarray) DO
            -- If the current element is not an ST_Point value, then
            -- raise an exception.
            IF ageometryarray[counter] IS NOT OF (ST_Point) THEN
                SIGNAL SQLSTATE '2FF08'
                SET MESSAGE_TEXT = 'element is not an ST_Point type';
            END IF;
            -- Cast the current element as an ST_Point and
            -- concatenate it to the end of apointarray.
            SET apointarray = apointarray || CAST(ageometryarray[counter] AS ST_Point);
            SET counter = counter + 1;
        END WHILE;
        -- Return an ST_Point array
        RETURN apointarray;
    END
```

Reply:
@martint Martin, are you saying that this code from spec suggests that empty points need to be included in the resulting Multi-Point geometry? If so, do you insist that Presto's implementation includes empty points as well or would it be acceptable to keep the current implementation that excludes empty points?

Reply:
The code seems to suggest empty points should be included, so our implementation should match that. Otherwise it’s an incompatibility with the spec.

Reply:
@martint I don't think it is possible to include an empty point into a multi-point geometry. WKT doesn't support that. ESRI library doesn't support that either.

Reply:
In that case, we should probably fail with an error instead of silently ignoring the values.

Have we looked at what other databases do in this case?

Reply:
@martint PostGIS doesn't have `ST_MultiPoint` function. The closest function I could find is `ST_Collect`: http://www.postgis.net/docs/ST_Collect.html

In PostGIS `ST_Collect(ARRAY[ST_GeometryFromText('POINT (1 2)'), ST_GeometryFromText('POINT EMPTY')])` returns `MULTIPOINT (1 2, EMPTY)` which is equal to `MULTIPOINT (1 2)` (e.g. `ST_Equals(ST_GeometryFromText('MULTIPOINT (1 2, EMPTY)'), ST_GeometryFromText('MULTIPOINT (1 2)'))` returns `true`.)

(BTW, `ST_GeometryFromText('MULTIPOINT (1 2, EMPTY)')` in Presto fails with `Invalid WKT: MULTIPOINT(1 2,EMPTY)`)

I think it is more user-friendly to skip empty points, then fail. PostGIS also allows empty points and returns geometries which are equivalent to what Presto would return.

Reply:
@martint *ping*

Reply:
It's preferable to be conservative and fail for invalid input. We can always relax constraints later, but making it stricter is a lot more complicated.